### PR TITLE
MergeDups refactor (all at once, rather than one-by-one)

### LINKED ISSUE #1072: Moving a sense from wordA to wordB also moves all the audio.
- Create entry with vernacular "WordA", 2 senses, and a recording.
- Create another entry with  vernacular "WordB" (must be the same or similar vernacular to the first).
- Go to Data Cleanup > Merge Duplicates.
- Move one of the two senses of WordA to WordB.
- Go to Data Cleanup > Review Entries, and the audio that was on WordA will now be on WordB.

Desired behavior: The audio should be still on WordA, not moved or copied. (The audio should be moved if and only if it is the last sense.)

### LINKED ISSUE #1078: Audio copying during merge inconsistent.
When a sense is moved from WordA to WordB, the audio should only move if WordA has no more senses and is to be deleted (see #1072). There is a check in place in `WordApiServices.cs > Merge()`, but it only sometimes works, due to the current implementation of merging--it may be that not enough information is sent to the backend after the first merge in a set of merges, when other related merges are updated.

### PR DESCRIPTION
**Pull request by imnasnainaec**
Rather than updating all remaining mergeWords after each is merged, prepare them all at once, then merge all, then delete all children from the frontier.

Resolves #1072
Resolves #1078

---
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1081)

### CODE REVIEW

**Comment by johnthagen** [comment]
___
*[Backend/Models/Word.cs, line 414 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbfXyX5X6D-i40rtmP:-MVbfXyX5X6D-i40rtmQ:bizka14) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Models/Word.cs#L414)):*
> ```Smalltalk
>         public Word Parent { get; set; }
> 
>         [BsonElement("children")]
> ```

Is this backwards compatible with old databases? e.g. were the old entries named `ChildrenWords`?

**Comment by johnthagen** [comment]
___
*[Backend/Models/Word.cs, line 414 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbfXyX5X6D-i40rtmP:-MVbfXyX5X6D-i40rtmQ:bizka14) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Models/Word.cs#L414)):*
> ```Smalltalk
>         public Word Parent { get; set; }
> 
>         [BsonElement("children")]
> ```

Is this backwards compatible with old databases? e.g. were the old entries named `ChildrenWords`?

**Comment by johnthagen** [comment]
___
*[Backend/Models/Word.cs, line 414 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbfXyX5X6D-i40rtmP:-MVbfXyX5X6D-i40rtmQ:bizka14) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Models/Word.cs#L414)):*
> ```Smalltalk
>         public Word Parent { get; set; }
> 
>         [BsonElement("children")]
> ```

Is this backwards compatible with old databases? e.g. were the old entries named `ChildrenWords`?

**Comment by johnthagen** [comment]
___
*[Backend/Models/Word.cs, line 430 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbgNQ74S2cvA4AZboI:-MVbgNQ74S2cvA4AZboJ:b-yv26uz) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Models/Word.cs#L430)):*
> ```Smalltalk
>         public string SrcWordId;
> 
>         [BsonElement("getAudio")]
> ```

Are there backwards compatibility concerns with this? Will old databases have this field?

**Comment by johnthagen** [comment]
___
*[Backend/Models/Word.cs, line 414 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbfXyX5X6D-i40rtmP:-MVbgguH0-lvE1Eztt1Y:b-i4f0ol) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Models/Word.cs#L414)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

Is this backwards compatible with old databases? e.g. were the old entries named `ChildrenWords`?
</blockquote></details>

Actually, we don't ever store these in the database, right? Why do we decorate them with `BsonElement`?

**Comment by johnthagen** [comment]
___
*[Backend/Services/WordApiServices.cs, line 161 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbh4HC4ongJri5HyEF:-MVbh4HC4ongJri5HyEG:b-7esed1) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Services/WordApiServices.cs#L161)):*
> ```Smalltalk
>             parent.Modified = "";
> 
>             if (wordsToAdd != null)
> ```

This can be simplified to `wordsToAdd?.Add(parent);` using `null` propagation.

**Comment by johnthagen** [comment]
___
*[Backend/Services/WordApiServices.cs, line 184 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbiFU8E5NxNd4YMUn4:-MVbiFU92R11BfmZ1QI-:b-1t88t9) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Services/WordApiServices.cs#L184)):*
> ```Smalltalk
>         {
>             var newWords = new List<Word>();
>             await Task.WhenAll(mergeWordsList.Select(m => MergePrepParent(projectId, m, newWords)));
> ```

Is there any way to collect all of the `parents` and construct `newWords` outside of `MergePrepParent`?

It was initially a little hard to follow the fact that `newWords` is passed in as a mutable list, and then `parent` is added to it. In a sense, `MergePrepParent` sort of has two return values (one of which is mutating `newWords`).

If there was a way to build `newWords` outside of the call using the return from `MergePrepParent` (which returns `parent`) I think it would make the function a little purer (fewer side effects) and easier to understand.

Something like:

```
// Psuedocode
newWords.Add(MergePrepParent(...))
```

**Comment by johnthagen** [comment]
___
*[Backend/Services/WordApiServices.cs, line 184 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbiFU8E5NxNd4YMUn4:-MVbiFU92R11BfmZ1QI-:b-1t88t9) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Services/WordApiServices.cs#L184)):*
> ```Smalltalk
>         {
>             var newWords = new List<Word>();
>             await Task.WhenAll(mergeWordsList.Select(m => MergePrepParent(projectId, m, newWords)));
> ```

Is there any way to collect all of the `parents` and construct `newWords` outside of `MergePrepParent`?

It was initially a little hard to follow the fact that `newWords` is passed in as a mutable list, and then `parent` is added to it. In a sense, `MergePrepParent` sort of has two return values (one of which is mutating `newWords`).

If there was a way to build `newWords` outside of the call using the return from `MergePrepParent` (which returns `parent`) I think it would make the function a little purer (fewer side effects) and easier to understand.

Something like:

```
// Psuedocode
newWords.Add(MergePrepParent(...))
```

**Comment by johnthagen** [comment]
___
*[Backend/Services/WordApiServices.cs, line 184 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbiFU8E5NxNd4YMUn4:-MVbiFU92R11BfmZ1QI-:b-1t88t9) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Services/WordApiServices.cs#L184)):*
> ```Smalltalk
>         {
>             var newWords = new List<Word>();
>             await Task.WhenAll(mergeWordsList.Select(m => MergePrepParent(projectId, m, newWords)));
> ```

Is there any way to collect all of the `parents` and construct `newWords` outside of `MergePrepParent`?

It was initially a little hard to follow the fact that `newWords` is passed in as a mutable list, and then `parent` is added to it. In a sense, `MergePrepParent` sort of has two return values (one of which is mutating `newWords`).

If there was a way to build `newWords` outside of the call using the return from `MergePrepParent` (which returns `parent`) I think it would make the function a little purer (fewer side effects) and easier to understand.

Something like:

```
// Psuedocode
newWords.Add(MergePrepParent(...))
```

**Comment by johnthagen** [comment]
___
*[Backend/Services/WordApiServices.cs, line 132 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbl4gM3rIm8pA2e2kg:-MVbl4gN1RnomhDaEBCH:b-i56z59) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Services/WordApiServices.cs#L132)):*
> ```Smalltalk
>         /// <returns> Word to add. </returns>
>         private async Task<Word> MergePrepParent(
>             string projectId, MergeWords mergeWords, List<Word>? wordsToAdd)
> ```

`wordsToAdd` is optional here, but we never use it without this parameter being set. Should we just require it to be present to remove a `null` check?

**Comment by johnthagen** [comment]
___
*[Backend/Services/WordApiServices.cs, line 185 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVblN_X8jo5B_Ai81S4:-MVblN_X8jo5B_Ai81S5:bphyjvo) ([raw file](https://github.com/sillsdev/thecombine/blob/57eab76de9145460455df37734aa32ee8037853b/Backend/Services/WordApiServices.cs#L185)):*
> ```Smalltalk
>                 for (var i = 0; i < currentChildWord.Senses.Count; i++)
>                 {
>                     switch (newChildWordState.SenseStates[i])
> ```

Out of curiosity, why were we able to remove this code? Was this latent functionality we weren't using in the application?

**Comment by johnthagen** [comment]
___
*[Backend/Services/WordRepository.cs, line 157 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVblxSA8ewoPXchIE2a:-MVblxSA8ewoPXchIE2b:b-1xdtdf) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Services/WordRepository.cs#L157)):*
> ```Smalltalk
> 
>         /// <summary> Removes <see cref="Word"/>s from the Frontier with specified wordIds and projectId </summary>
>         /// <returns> A long: number of words deleted </returns>
> ```

The "long" is part of the function signature, so I think we could simplify this to:

```
Number of words deleted
```

**Review by imnasnainaec — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1081)* status: 0 of 12 files reviewed, 7 unresolved discussions (waiting on @imnasnainaec, @jasonleenaylor, and @johnthagen)

___
*[Backend/Controllers/WordController.cs, line 213 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbe8T1-Mkcf68VZXwr:-MVbj-bOAygLV8CcIz1j:blsoqpy) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Controllers/WordController.cs#L213)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

Typo: I meant "blanket-suppressing"
</blockquote></details>

Indeed. Jason made steps toward logging this kind of thing in #1074 and I think we are generally interested in improving.
___
*[Backend/Models/Word.cs, line 414 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbfXyX5X6D-i40rtmP:-MVblm7g8i78K96G7tLV:bmnd3zm) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Models/Word.cs#L414)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

Actually, we don't ever store these in the database, right? Why do we decorate them with `BsonElement`?
</blockquote></details>

Good point. Removed.
___
*[Backend/Models/Word.cs, line 430 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbgNQ74S2cvA4AZboI:-MVblo5t8vbY97GJK0Zi:b-896fix) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Models/Word.cs#L430)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

Are there backwards compatibility concerns with this? Will old databases have this field?
</blockquote></details>

Done.

**Comment by johnthagen** [comment]
___
*[Backend.Tests/Controllers/WordControllerTests.cs, line 197 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbmNay3dmv_-Lm0M5h:-MVbmNay3dmv_-Lm0M5i:b-rpucic) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend.Tests/Controllers/WordControllerTests.cs#L197)):*
> ```Smalltalk
>             };
> 
>             var newWords = _wordService.Merge(_projId, new List<MergeWords>() { mergeObject }).Result;
> ```

Can remove the `()`, that is not needed when using `{ }` initialization.

**Review by imnasnainaec — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1081)* status: 0 of 12 files reviewed, 6 unresolved discussions (waiting on @imnasnainaec, @jasonleenaylor, and @johnthagen)

___
*[Backend/Services/WordApiServices.cs, line 184 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbiFU8E5NxNd4YMUn4:-MVbmGy_6jJnQVsWVUXj:bj4fvy0) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Services/WordApiServices.cs#L184)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

Is there any way to collect all of the `parents` and construct `newWords` outside of `MergePrepParent`?

It was initially a little hard to follow the fact that `newWords` is passed in as a mutable list, and then `parent` is added to it. In a sense, `MergePrepParent` sort of has two return values (one of which is mutating `newWords`).

If there was a way to build `newWords` outside of the call using the return from `MergePrepParent` (which returns `parent`) I think it would make the function a little purer (fewer side effects) and easier to understand.

Something like:

```
// Psuedocode
newWords.Add(MergePrepParent(...))
```
</blockquote></details>

How's this?
___
*[Backend/Services/WordApiServices.cs, line 185 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVblN_X8jo5B_Ai81S4:-MVbmW7q-HvCpGOHe1J4:b6ltjvm) ([raw file](https://github.com/sillsdev/thecombine/blob/57eab76de9145460455df37734aa32ee8037853b/Backend/Services/WordApiServices.cs#L185)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

Out of curiosity, why were we able to remove this code? Was this latent functionality we weren't using in the application?
</blockquote></details>

Before this refactor, children of one merge were being updated in both the backend and the frontend. By computing all the parents at once, we don't have to modify the children. We can just delete them all from the frontier once the parents are constructed.

**Review by imnasnainaec — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1081)* status: 0 of 12 files reviewed, 6 unresolved discussions (waiting on @imnasnainaec, @jasonleenaylor, and @johnthagen)

___
*[Backend/Services/WordRepository.cs, line 157 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVblxSA8ewoPXchIE2a:-MVbncuH2nFFTeZW_kyL:b-896fix) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Services/WordRepository.cs#L157)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

The "long" is part of the function signature, so I think we could simplify this to:

```
Number of words deleted
```
</blockquote></details>

Done.
___
*[Backend.Tests/Controllers/WordControllerTests.cs, line 197 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbmNay3dmv_-Lm0M5h:-MVbndgRDv0WX09TdJRQ:b-896fix) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend.Tests/Controllers/WordControllerTests.cs#L197)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

Can remove the `()`, that is not needed when using `{ }` initialization.
</blockquote></details>

Done.

**Comment by johnthagen** [comment]
___
*[Backend.Tests/Controllers/WordControllerTests.cs, line 219 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbnY4m-bFR55G7Kd0w:-MVbnY4m-bFR55G7Kd0x:b-kjdclf) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend.Tests/Controllers/WordControllerTests.cs#L219)):*
> ```Smalltalk
>             var mergeWords = new MergeWords { Parent = RandomWord() };
>             const int numberOfChildren = 3;
>             for (int i = 0; i < numberOfChildren; i++)
> ```

Since we aren't using `i` for anything, we could express this more clearly as:

```
foreach (var _ in Enumerable.Range(0, numberOfChildren))
```

**Review by imnasnainaec — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1081)* status: 0 of 12 files reviewed, 7 unresolved discussions (waiting on @imnasnainaec, @jasonleenaylor, and @johnthagen)

___
*[Backend.Tests/Controllers/WordControllerTests.cs, line 219 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbnY4m-bFR55G7Kd0w:-MVbo2R45fcVZsmRjUYH:b-896fix) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend.Tests/Controllers/WordControllerTests.cs#L219)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

Since we aren't using `i` for anything, we could express this more clearly as:

```
foreach (var _ in Enumerable.Range(0, numberOfChildren))
```
</blockquote></details>

Done.

**Comment by johnthagen** [comment]
___
*[Backend.Tests/Controllers/WordControllerTests.cs, line 248 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbnqsSE-r6ORUmiLu0:-MVbnqsSE-r6ORUmiLu1:b-brna78) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend.Tests/Controllers/WordControllerTests.cs#L248)):*
> ```Smalltalk
>             var newWords = _wordService.Merge(_projId, mergeWordsList).Result;
> 
>             Assert.That(newWords, Has.Count.EqualTo(2));
> ```

Could replace usages of `2` with `mergeWordsList.Count`?

Though *technically* that list could get mutated in `.Merge` so maybe leaving `2` is a little more "correct".

**Comment by johnthagen** [comment]
___
*[Backend.Tests/Controllers/WordControllerTests.cs, line 245 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVboDjn3hzf74uJ_P9o:-MVboDjn3hzf74uJ_P9p:b-xu7f7o) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend.Tests/Controllers/WordControllerTests.cs#L245)):*
> ```Smalltalk
>             var mergeWordsA = new MergeWords { Parent = RandomWord() };
>             var mergeWordsB = new MergeWords { Parent = RandomWord() };
>             var mergeWordsList = new List<MergeWords>() { mergeWordsA, mergeWordsB };
> ```

`()` not needed here.

**Review by imnasnainaec — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1081)* status: 0 of 14 files reviewed, 8 unresolved discussions (waiting on @imnasnainaec, @jasonleenaylor, and @johnthagen)

___
*[Backend.Tests/Controllers/WordControllerTests.cs, line 245 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVboDjn3hzf74uJ_P9o:-MVbp80c4XrnYNmn3SZ2:b-896fix) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend.Tests/Controllers/WordControllerTests.cs#L245)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

`()` not needed here.
</blockquote></details>

Done.
___
*[Backend.Tests/Controllers/WordControllerTests.cs, line 248 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbnqsSE-r6ORUmiLu0:-MVbp8_G9NeZr5MW8yG5:b-myynsy) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend.Tests/Controllers/WordControllerTests.cs#L248)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

Could replace usages of `2` with `mergeWordsList.Count`?

Though *technically* that list could get mutated in `.Merge` so maybe leaving `2` is a little more "correct".
</blockquote></details>

I'm inclined to the explicit 2, due to the mutability.

**Comment by johnthagen** [comment]
___
*[src/goals/MergeDupGoal/MergeDupStep/MergeDupStepActions.tsx, line 158 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbpkdR7mrwuZmThYqX:-MVbpkdR7mrwuZmThYqY:bdbssy6) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/src/goals/MergeDupGoal/MergeDupStep/MergeDupStepActions.tsx#L158)):*
> ```XML
> type SenseWithState = TreeDataSense & { state: State };
> 
> function getMergeWords(
> ```

This function is quite hefty. Perhaps we should consider breaking it down into a couple private functions?

**Comment by johnthagen** [comment]
___
*[src/goals/MergeDupGoal/MergeDupStep/MergeDupStepActions.tsx, line 161 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbpv3Q6uGO-PX0f-E-:-MVbpv3R3-w7j0J8WQjD:bt9lnj0) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/src/goals/MergeDupGoal/MergeDupStep/MergeDupStepActions.tsx#L161)):*
> ```XML
>   wordId: string,
>   getState: () => StoreState
> ): MergeWords | undefined {
> ```

Might be worth documenting at the top what it means / when it can happen that `undefined` is returned.

**Comment by johnthagen** [comment]
___
*[Backend/Services/WordApiServices.cs, line 184 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbiFU8E5NxNd4YMUn4:-MVbqnPMBKowXUsDumCg:b-dbdhei) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Services/WordApiServices.cs#L184)):*
<details><summary><i>Previously, imnasnainaec (D. Ror.) wrote…</i></summary><blockquote>

How's this?
</blockquote></details>

That looks much better!

I think for local variables it should be `task` rather than `Task`, but if this works, to me this is much clearer and keeps `MergePrepParent` a lot simpler.

**Review by imnasnainaec — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1081)* status: 0 of 14 files reviewed, 8 unresolved discussions (waiting on @imnasnainaec, @jasonleenaylor, and @johnthagen)

___
*[src/goals/MergeDupGoal/MergeDupStep/MergeDupStepActions.tsx, line 158 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbpkdR7mrwuZmThYqX:-MVbqhMt0PGdnX1RECn3:b-klx6xn) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/src/goals/MergeDupGoal/MergeDupStep/MergeDupStepActions.tsx#L158)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

This function is quite hefty. Perhaps we should consider breaking it down into a couple private functions?
</blockquote></details>

I expect to do more refactoring of the frontend soon, and will look at further simplifying/deconstructing this function then.

**Comment by johnthagen** [comment]
___
*[Backend.Tests/Controllers/WordControllerTests.cs, line 248 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbnqsSE-r6ORUmiLu0:-MVbrE8C2Bv7h1wkPjZP:b-pbdqz3) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend.Tests/Controllers/WordControllerTests.cs#L248)):*
<details><summary><i>Previously, imnasnainaec (D. Ror.) wrote…</i></summary><blockquote>

I'm inclined to the explicit 2, due to the mutability.
</blockquote></details>

Agreed.

**Comment by johnthagen** [comment]
___
*[Backend/Services/WordApiServices.cs, line 161 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbh4HC4ongJri5HyEF:-MVbrL3g6CnY69yqdnc5:b-896fix) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/Backend/Services/WordApiServices.cs#L161)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

This can be simplified to `wordsToAdd?.Add(parent);` using `null` propagation.
</blockquote></details>

Done.

**Comment by johnthagen** [comment]
___
*[Backend/Services/WordApiServices.cs, line 180 at r3](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbrSSJAchTDUnKy8rm:-MVbrSSJAchTDUnKy8rn:b-48xjbh) ([raw file](https://github.com/sillsdev/thecombine/blob/2f97885a1c36d49584dd86b5e6e482969bbb8f06/Backend/Services/WordApiServices.cs#L180)):*
> ```Smalltalk
>             var newWords = new List<Word>();
>             await Task.WhenAll(mergeWordsList.Select(m => MergePrepParent(projectId, m)
>                                              .ContinueWith(Task => newWords.Add(Task.Result))));
> ```

`Task` -> `task` (local variable naming style)

**Review by johnthagen — CHANGES_REQUESTED** [review]
> Reviewed 7 of 12 files at r1, 5 of 5 files at r2, 6 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1081)* status: all files reviewed, 2 unresolved discussions (waiting on @imnasnainaec and @jasonleenaylor)

**Comment by johnthagen** [comment]
___
*[Backend/Services/WordApiServices.cs, line 180 at r3](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbrSSJAchTDUnKy8rm:-MVbsBFT3BnS0gkwd7_n:b-896fix) ([raw file](https://github.com/sillsdev/thecombine/blob/2f97885a1c36d49584dd86b5e6e482969bbb8f06/Backend/Services/WordApiServices.cs#L180)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

`Task` -> `task` (local variable naming style)
</blockquote></details>

Done.

**Review by imnasnainaec — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1081)* status: 12 of 14 files reviewed, 1 unresolved discussion (waiting on @jasonleenaylor and @johnthagen)

___
*[src/goals/MergeDupGoal/MergeDupStep/MergeDupStepActions.tsx, line 161 at r2](https://reviewable.io/reviews/sillsdev/thecombine/1081#-MVbpv3Q6uGO-PX0f-E-:-MVbsFB48db3WJFXd8YW:b-896fix) ([raw file](https://github.com/sillsdev/thecombine/blob/ff90b7daa21d8d6a2306a7fad43de2b9ea3b31b4/src/goals/MergeDupGoal/MergeDupStep/MergeDupStepActions.tsx#L161)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

Might be worth documenting at the top what it means / when it can happen that `undefined` is returned.
</blockquote></details>

Done.

**Review by johnthagen — APPROVED** [review]
> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1081)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @jasonleenaylor)
